db/inlineskiplist.h

```diff
@@ -673,4 +883,41 @@ bool InlineSkipList<Comparator>::Contains(const char* key) const {
   }
 }
 
+template <class Comparator>
+void InlineSkipList<Comparator>::TEST_Validate() const {
+  // Interate over all levels at the same time, and verify nodes appear in
+  // the right order, and nodes appear in upper level also appear in lower
+  // levels.
+  Node* nodes[kMaxPossibleHeight];
+  int max_height = GetMaxHeight();
+  for (int i = 0; i < max_height; i++) {
+    nodes[i] = head_;
+  }
+  while (nodes[0] != nullptr) {
+    nodes[0] = nodes[0]->Next(0);
+    if (nodes[0] == nullptr) {
+      break;
+    }
+    int i = 1;
+    while (i < max_height) {
+      Node* next = nodes[i]->Next(i);
+      if (next == nullptr) {
+        break;
+      }
+      auto cmp = compare_(nodes[0]->Key(), next->Key());
+      assert(cmp <= 0);
+      if (cmp == 0) {
+        assert(next == nodes[0]);
+        nodes[i] = next;
+      } else {
+        break;
+      }
+      i++;
+    }
+  }
+  for (int i = 1; i < max_height; i++) {
+    assert(nodes[i]->Next(i) == nullptr);
+  }
```

Comment:
I think you also need to verify all the levels in all the nodes are used. Otherwise, you may get something like this:

```

+-+             +-+          ++
| |             | |          ||
| |             | |          ||
| |             | |          ||
| |             | |          ||
| +------------------------> |------>
| |             | |          ||
| |             | |          ||
| +-----------> | +------------------>
| |             | |          ||
| |             | |          ||
| | +---------> | | +------> ||
+-+             +-+          ++

```
which will not get you the correct result.